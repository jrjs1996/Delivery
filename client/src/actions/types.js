export const FETCH_CUSTOMERS = 'FETCH_CUSTOMERS';
export const ADD_CUSTOMER = 'ADD_CUSTOMER';

export const ADD_ORDER = 'ADD_ORDER';
export const FETCH_ORDERS = 'FETCH_ORDER';
export const COMPLETE_ORDER = 'COMPLETE_ORDER';
export const GET_CUSTOMER = 'GET_CUSTOMER';

export const LOGIN_ADMIN = 'LOGIN_ADMIN';
export const FETCH_ADMINS = 'FETCH_ADMINS';

export const FETCH_MENU = 'FETCH_MENU';
export const ADD_MENU_ITEM = 'ADD_MENU_ITEM';
export const UPDATE_MENU_ITEM = 'UPDATE_MENU_ITEM';
export const DELETE_MENU_ITEM = 'DELETE_MENU_ITEM';
