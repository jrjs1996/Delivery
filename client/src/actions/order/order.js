import axios from 'axios';
import {
  ADD_ORDER,
  ADD_TO_CURRENT_ORDER,
  DELETE_CURRENT_ORDER,
  ERROR,
  FETCH_ORDERS,
  REMOVE_FROM_CURRENT_ORDER,
  UPDATE_ORDER,
} from '../types';
import { setAuthHeader } from '../../utils/token';

export const addToCurrentOrder = item => (dispatch) => {
  dispatch({
    type: ADD_TO_CURRENT_ORDER,
    payload: item,
  });
};

export const createOrder = (currentOrder, orderInfo = {}) => async (dispatch) => {
  try {
    const postData = Object.assign(orderInfo);
    postData.items = {};
    Object.keys(currentOrder.items).forEach((i) => {
      postData.items[i] = currentOrder.items[i].count;
    });
    const res = await axios.post('/api/orders/', postData);
    dispatch({
      type: ADD_ORDER,
      payload: res.data,
    });
  } catch (error) {
    console.log(error);
    dispatch({
      type: ERROR,
      payload: 'Unable to create order',
    });
  }
};

export const deleteCurrentOrder = () => (dispatch) => {
  dispatch({
    type: DELETE_CURRENT_ORDER,
  });
};

export const fetchOpenOrders = () => async (dispatch) => {
  try {
    const res = await axios.get('/api/orders/?toStage=3', setAuthHeader());
    dispatch({
      type: FETCH_ORDERS,
      payload: res.data,
    });
  } catch (error) {
    dispatch({
      type: ERROR,
      payload: 'Unable to fetch orders',
    });
  }
};

export const fetchOrders = () => async (dispatch) => {
  try {
    const res = await axios.get('/api/orders/', setAuthHeader());
    dispatch({
      type: FETCH_ORDERS,
      payload: res.data,
    });
  } catch (error) {
    dispatch({
      type: ERROR,
      payload: 'Unable to fetch orders',
    });
  }
};

export const removeFromCurrentOrder = menuNumber => (dispatch) => {
  dispatch({
    type: REMOVE_FROM_CURRENT_ORDER,
    payload: menuNumber,
  });
};

export const updateOrder = (id, putData) => async (dispatch) => {
  try {
    await axios.put(`/api/orders/${id}`, putData, setAuthHeader());
    const payload = putData;
    payload._id = id;
    dispatch({
      type: UPDATE_ORDER,
      payload,
    });
  } catch (error) {
    dispatch({
      type: ERROR,
      payload: 'Unable to update order.',
    });
  }
};
