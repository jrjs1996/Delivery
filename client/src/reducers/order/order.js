import {
  FETCH_ORDERS, ADD_ORDER, UPDATE_ORDER, ADD_TO_CURRENT_ORDER, REMOVE_FROM_CURRENT_ORDER,
} from '../../actions/types';
import { createItems, insertItem, updateItem } from '../utils/utils';

const initialState = {
  items: {},
  currentOrder: {},
};

const addToCurrentOrder = (currentOrder, item) => {
  const newOrder = { ...currentOrder };
  // If the item is already in the order add 1 to count
  if (newOrder[item._id]) {
    newOrder[item._id].count += 1;
    return newOrder;
  }
  // Otherwise add it to the order
  newOrder[item._id] = {
    item,
    count: 1,
  };
  return newOrder;
};

const removeFromCurrentOrder = (currentOrder, itemId) => {
  const newOrder = { ...currentOrder };
  if (!newOrder[itemId]) return newOrder;

  newOrder[itemId].count -= 1;
  // If count is less than 1 remove it from the order
  if (newOrder[itemId].count < 1) {
    delete newOrder[itemId];
  }
  return newOrder;
};

export default function (state = initialState, action) {
  switch (action.type) {
    case ADD_ORDER:
      return {
        ...state,
        items: insertItem(state.items, action.payload, '_id'),
      };
    case FETCH_ORDERS:
      return {
        ...state,
        items: createItems(action.payload, '_id'),
      };
    case ADD_TO_CURRENT_ORDER:
      return {
        ...state,
        currentOrder: addToCurrentOrder(state.currentOrder, action.payload),
      };
    case UPDATE_ORDER:
      return {
        ...state,
        items: updateItem(state.items, action.payload, '_id'),
      };
    case REMOVE_FROM_CURRENT_ORDER:
      return {
        ...state,
        currentOrder: removeFromCurrentOrder(state.currentOrder, action.payload),
      };
    default:
      return state;
  }
}
