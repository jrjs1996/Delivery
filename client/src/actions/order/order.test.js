import configureStore from 'redux-mock-store';
import * as orderActions from './order';
import * as actionTypes from '../types';
import mockAxios from '../../../__mocks__/axios';

const mockStore = configureStore();
const store = mockStore();

describe('Create order', () => {
  beforeEach(() => {
    store.clearActions();
  });

  it('Dispatches ADD_ORDER with payload of response data on success', async () => {
    await orderActions.createOrder({ mockPostData: 'MockPostData' })(store.dispatch);
    expect(store.getActions()).toMatchSnapshot();
  });

  it('Dispatches ERROR on error', async () => {
    mockAxios.giveError(mockAxios.post);
    await orderActions.createOrder({ mockPostData: 'MockPostData' })(store.dispatch);
    expect(store.getActions()).toMatchSnapshot();
  });
});

describe('Delete Current Order', () => {
  beforeEach(() => {
    store.clearActions();
  });

  it('Dispatches DELETE_CURRENT_ORDER', () => {
    orderActions.deleteCurrentOrder()(store.dispatch);
    expect(store.getActions()).toMatchSnapshot();
  });
});

describe('Fetch open orders', () => {
  beforeEach(() => {
    store.clearActions();
  });

  it('Dispatches FETCH_ORDERS with response data on success', async () => {
    await orderActions.fetchOpenOrders()(store.dispatch);
    expect(store.getActions()).toMatchSnapshot();
  });

  it('Dispatches ERROR on error', async () => {
    mockAxios.giveError(mockAxios.get);
    await orderActions.fetchOpenOrders()(store.dispatch);
    expect(store.getActions()).toMatchSnapshot();
  });
});

describe('Fetch orders', () => {
  beforeEach(() => {
    store.clearActions();
  });

  it('Dispatches FETCH_ORDERS with response data on success', async () => {
    await orderActions.fetchOrders()(store.dispatch);
    expect(store.getActions()).toMatchSnapshot();
  });

  it('Dispatches ERROR on error', async () => {
    mockAxios.giveError(mockAxios.get);
    await orderActions.fetchOrders()(store.dispatch);
    expect(store.getActions()).toMatchSnapshot();
  });
});

describe('Update order', () => {
  beforeEach(() => {
    store.clearActions();
  });

  it('Dispatches UPDATE_ORDER with given id and putdata on success', async () => {
    await orderActions.updateOrder(1234, { mockPutData: 'MockPutData' })(store.dispatch);
    expect(store.getActions()).toMatchSnapshot();
  });

  it('Dispatches ERROR on error', async () => {
    mockAxios.giveError(mockAxios.put);
    await orderActions.updateOrder({ mockPutData: 'MockPutData' })(store.dispatch);
    expect(store.getActions()).toMatchSnapshot();
  });
});
