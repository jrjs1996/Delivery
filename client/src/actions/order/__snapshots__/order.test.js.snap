// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create order Dispatches ADD_ORDER with payload of response data on success 1`] = `
Array [
  Object {
    "payload": Object {
      "mock": "MockData",
    },
    "type": "ADD_ORDER",
  },
]
`;

exports[`Create order Dispatches ERROR on error 1`] = `
Array [
  Object {
    "payload": "Unable to create order",
    "type": "ERROR",
  },
]
`;

exports[`Fetch open orders Dispatches ERROR on error 1`] = `
Array [
  Object {
    "payload": "Unable to fetch orders",
    "type": "ERROR",
  },
]
`;

exports[`Fetch open orders Dispatches FETCH_ORDERS with response data on success 1`] = `
Array [
  Object {
    "payload": Object {
      "mock": "MockData",
    },
    "type": "FETCH_ORDER",
  },
]
`;

exports[`Fetch orders Dispatches ERROR on error 1`] = `
Array [
  Object {
    "payload": "Unable to fetch orders",
    "type": "ERROR",
  },
]
`;

exports[`Fetch orders Dispatches FETCH_ORDERS with response data on success 1`] = `
Array [
  Object {
    "payload": Object {
      "mock": "MockData",
    },
    "type": "FETCH_ORDER",
  },
]
`;

exports[`Update order Dispatches ERROR on error 1`] = `
Array [
  Object {
    "payload": "Unable to update order.",
    "type": "ERROR",
  },
]
`;

exports[`Update order Dispatches UPDATE_ORDER with given id and putdata on success 1`] = `
Array [
  Object {
    "payload": Object {
      "_id": 1234,
      "mockPutData": "MockPutData",
    },
    "type": "UPDATE_ORDER",
  },
]
`;
